heartbeat/storage-mon.in

```diff
@@ -173,77 +205,163 @@ storage-mon_init() {
 	fi
 }
 
-storage-mon_validate() {
-	storage-mon_init
-
-	# Is the state directory writable?
-	state_dir=$(dirname "$OCF_RESKEY_state_file")
-	touch "$state_dir/$$"
-	if [ $? -ne 0 ]; then
-		return $OCF_ERR_CONFIGURED
-	fi
-	rm "$state_dir/$$"
-
-	return $OCF_SUCCESS
-}
-
 storage-mon_monitor() {
-	storage-mon_init
+	if [ -z "$OCF_RESKEY_daemonize" ]; then
+		storage-mon_init
 
-	# Monitor _MUST!_ differentiate correctly between running
-	# (SUCCESS), failed (ERROR) or _cleanly_ stopped (NOT RUNNING).
-	# That is THREE states, not just yes/no.
+		# Monitor _MUST!_ differentiate correctly between running
+		# (SUCCESS), failed (ERROR) or _cleanly_ stopped (NOT RUNNING).
+		# That is THREE states, not just yes/no.
 
-	if [ ! -f "${OCF_RESKEY_state_file}" ]; then
-		return $OCF_NOT_RUNNING
-	fi
+		if [ ! -f "${OCF_RESKEY_state_file}" ]; then
+			return $OCF_NOT_RUNNING
+		fi
 
-	# generate command line
-	cmdline=""
-	for DRIVE in ${OCF_RESKEY_drives}; do
-		cmdline="$cmdline --device $DRIVE --score 1"
-	done
-	cmdline="$cmdline --timeout ${OCF_RESKEY_io_timeout}"
-	if [ -n "${OCF_RESKEY_inject_errors}" ]; then
-		cmdline="$cmdline --inject-errors-percent ${OCF_RESKEY_inject_errors}"
-	fi
-	$STORAGEMON $cmdline
-	if [ $? -ne 0 ]; then
-		status="red"
+		# generate command line
+		cmdline=""
+		for DRIVE in ${OCF_RESKEY_drives}; do
+			cmdline="$cmdline --device $DRIVE --score 1"
+		done
+		cmdline="$cmdline --timeout ${OCF_RESKEY_io_timeout}"
+		if [ -n "${OCF_RESKEY_inject_errors}" ]; then
+			cmdline="$cmdline --inject-errors-percent ${OCF_RESKEY_inject_errors}"
+		fi
+		$STORAGEMON $cmdline
+		if [ $? -ne 0 ]; then
+			status="red"
+		else
+			status="green"
+		fi
+
+		"$ATTRDUP" -n ${ATTRNAME} -U "$status" -d "5s"
+		return $OCF_SUCCESS
 	else
-		status="green"
-	fi
+		ocf_pidfile_status "${PIDFILE}" > /dev/null 2>&1
+		case "$?" in
+			0)	rc=$OCF_SUCCESS;;
+			1|2)	rc=$OCF_NOT_RUNNING;;
+			*)	rc=$OCF_ERR_GENERIC;;
+		esac
+
+		if [ $rc -ne $OCF_SUCCESS ]; then
+			return "$rc"
+		fi
+		if [ "$1" = "pid_check_only" ]; then
+			return "$rc"
+		fi
 
-	"$ATTRDUP" -n "#health-${OCF_RESOURCE_INSTANCE}" -U "$status" -d "5s"
-	return $OCF_SUCCESS
+		# generate client command line
+		cmdline=""
+		cmdline="$cmdline --client --attrname ${ATTRNAME}"
+		while :
+		do
+			# 0			: Normal.
+			# greater than 0	: monitoring error.
+			# 255(-1)		: communication system error.
+			# 254(-2)		: Not all checks completed for first device in daemon mode.
+			$STORAGEMON $cmdline
+			rc=$?
+			case "$rc" in
+				254|255)
+					# If there is a communication error or the initial check of all devices has not been completed, 
+					# it will loop and try to reconnect.
+					# When everything ends with a communication error during monitor, a monitor timeout occurs.
+					ocf_log debug "client monitor error : $rc"
```

Comment:
Same comment as above applies for $STORAGEMON, $ATTRDUP (+ anything I missed of the same category ;-) ) here in the daemonize section.
Triggering a monitoring-timeout for some exit-codes in the end is of course fine as well.

Reply:
The first atttrd_updater call in the monitor is the original storage-mon RA code, not considering failures.
I didn't consider failures in the daemon/client mode side as well, but as you commented, I think it's correct to consider failures of attrd_updater.
Let me think a little more about the case when attrd_updater fails.

Reply:
Yep I know it is actually not in the context of this PR. Just had issues with it being like that recently.  OTOH with the new stuff we anyway have to go over what could possibly go wrong again and how we'd like to react. To quite some extent handling of the new client is anyway quite comparable to handling in all-in-one mode.

Reply:
Hi @HideoYamauchi !

I guess looping in case of issues setting the attribute and having pacemaker deal with
the possible timeout is a good idea.

I'm not sure if I'm getting what you want to say with the last paragraph.
If it is about finding cases where we could speed up failing I'm with you that this is good as well.
All in all the bullet-proof approach to retry for everything unknown and if we encounter an explicit
failure scenario we want to speed up then explicitly handle that.

Regards,
Klaus

Reply:
Hi @wenningerk,

Okay!

On the storage-mon RA side, let red attribute handle failure immediately, and for green, let's handle failure of attrd_updter in a loop.

```
                while :
                do
                        "$ATTRDUP" -n ${ATTRNAME} -U "$status" -d "5s"
                        rc=$?
                        if [ $rc -eq 0 ]; then
                                break
                        fi

                        ocf_log debug "${status} attribute by attrd_updater failed"

                        if [ $status = "red" ]; then
                                # If the attrd_updater fails with the red attribute, return an error to let pacemaker handle the failure immediately.
                                return $OCF_ERR_GENERIC
                        fi
                done
```

For this correspondence, I found that one point of RA needs to be modified.
I will push this later.

Best Regards,
Hideo Yamauchi.